Documentation/planning/multi-sdk-band-support.md

```diff
@@ -0,0 +1,253 @@
+# Source-Build Support for Multiple SDKs
+
+This document serves as the design planning document for how .NET source build will support multiple SDK bands. More generally, this support could be described as "partial VMR support". This is the implementation plan for https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Managing-SDK-Bands.md.
+
+## Overview and General Approach
+
+The general approach to supporting more than one SDK band is to **not** view the build of additional SDK bands as being any different from any other source build. Each build of the product combines a set of inputs (previously source-built artifacts + source) to produce a set of outputs that can ship to customers. When building two completely different major versions of .NET, the set of inputs is different (e.g. different source, 6.0 SDK vs. 7.0 SDK). Some of the previously source-built artifacts come from .NET, some may come from the source-built package ecosystem (e.g. icu or clang/llvm). If a source built product repo were to eliminate part of the input sources, instead relying on a prebuilt binary, that prebuilt binary would need to come from a set of source-built artifacts.
+
+The primary restriction on building multiple SDK bands is that they must not differ in the shared runtimes. For N bands, there must be a single runtime artifact. Thus, we can view a source-build of N SDK bands as:
+- As a single build of one runtime, producing a set of source-built outputs and **then**
+- Building N SDK bands **using**
+- The *outputs* of the runtime build combined with additional previously source-built artifacts (e.g. a matching band SDK, required tools and libraries) and source, to produce N SDK bands that meet source-build requirements.
+
+The runtime build requires a 1xx band SDK to build **and** redistributes functionality from tooling that ships in the 1xx band (e.g. it has dependencies on roslyn and roslyn-analyzers). Furthermore, most changes to a .NET major version occur when only one SDK band exists. Therefore it makes practical sense at this point to include the source of the 1xx SDK and the shared runtime in the same VMR branch, and to produce all artifacts related to 1xx, including the runtime, in its build. Newer bands will be VMRs that contain **only** the components that **differ** from the 1xx band. Only those components that differ. When building the newer band, the outputs of the 1xx build would be fed into the newer band build, in a similar manner as previously source built artifacts are today. If the source for a component does not exist in a given VMR, it is simply not built, and obviously no outputs would be produced for that component. Thus any downstream dependencies would be forced to use the versions provided by input source built artifacts, typically from the 1xx source build. If no such version exists, then the input would fail to be found or restored from online sources and be reported as a prebuilt.
```

Comment:
We should be explicit by saying something like 'dotnet/dotnet VMR'. I think of VMRs as a concept, not as a specific instance. I think this is pretty important. I can fork, change, etc. a repo. It may relate to the original repo, but it's not the same repo any longer. But it is a repo.

Reply:
I understand this but we need to be very careful about the terminology here to make this easily understandable to people who are not participating on the project daily. People will (and already are) using the term VMR primarily for the dotnet/dotnet repo instance itself, which is natural as it's already a repo. I'm concerned that overloading the term "repo" by calling a concrete projection (such as the 1xx sdk band) a repo as well might cause lot of misunderstandings in the future. I also don't have much better suggestions (perhaps "projection") for better naming here, unfortunatelly.

Reply:
Alright, I think I clarified this by specifying "dotnet VMR" and "dotnet VMR branch" where applicable.